### PR TITLE
Feat footnote

### PR DESCRIPTION
**Pull request by Jocs**
| Q                | A
| ---------------- | ---
| New feature?     | yes
| Fixed tickets    | #193 #1568 
| License          | MIT

### Description

屏幕快照 2019-10-27 下午7 06 05

- [x] Create footnote from `footnote identifier`
- [x] Jump to footnote from `footnote identifier`
- [x] Jump to `footnote identifier` from `footnote`
- [x] Handle `backspace` in footnote
- [x] Export markdown with footnote
- [x] Import markdown with footnote
- [x] Footnote setting option in setting window

---
This change is [Reviewable](https://reviewable.io/reviews/marktext/marktext/1546)

### CODE REVIEW

File: src/muya/lib/parser/index.js [hunk 1]
```diff
@@ -223,6 +223,27 @@ const tokenizerFac = (src, beginRules, inlineRules, pos = 0, top, labels) => {
         continue
       }
     }
+    // footnote identifier
```
Comment (fxha) [inline src/muya/lib/parser/index.js]: Newline is recommended.

File: src/muya/lib/parser/marked/utils.js [hunk 2]
```diff
@@ -2,6 +2,10 @@
  * Helpers
  */
 
+let i = 0
```
Comment (fxha) [inline src/muya/lib/parser/marked/utils.js]: Please use a different variable name like `uniqueIdCounter`.

File: src/muya/lib/parser/marked/lexer.js [hunk 3]
```diff
@@ -28,15 +29,39 @@ Lexer.prototype.lex = function (src) {
     .replace(/\r\n|\r/g, '\n')
     .replace(/\t/g, '    ')
   this.checkFrontmatter = true
-  return this.token(src, true)
+  this.token(src, true)
+  // Move footnote token to the end of tokens.
```
[the remark this answers is not in the document]
Comment (Jocs) [reply src/muya/lib/parser/marked/lexer.js]: Export html/PDF also need to set footnotes at the end of document.

**Comment by Jocs** [comment]
>Table inside footnote block doesn't work - no design and seems buggy.

It seems work.
屏幕快照 2019-10-30 下午12 45 30

**Comment by Jocs** [comment]
>Cannot create paragraph behind a code/HTML/math/diagram block in footnote block and context menu create paragraph after just insert a new paragraph behind the footnote block.

Any ideas to create paragraph bellow these blocks? we can only create new paragraph bellow these blocks in source code mode. BTW, maybe not much users use multiple blocks footnote :smile:

**Comment by Jocs** [comment]
>Please remove debug code

Fixed

>[^longnote] is not auto numbered in exported document but in footnotes.

Fixed

>Spell checking is enabled on ag-footnote-input (footnote identifer in display block)

Fixed

>Footnote display box should be also creatable with Enter (--> [^foo]:<enter>)

Fixed

>Table inside footnote block doesn't work - no design and seems buggy.

Need discuss.

>Cannot create paragraph behind a code/HTML/math/diagram block in footnote block and context menu create paragraph after just insert a new paragraph behind the footnote block.

Need discuss.

File: src/muya/lib/parser/marked/lexer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fxha) [reply src/muya/lib/parser/marked/lexer.js]: That's why add an option to allow to disable it.

**Comment by fxha** [comment]
Related block content: I think the use case is very limited but a footnote block is the same as a list or quote block, right? We could keep the issue with the new paragraph in mind but should support block elements inside.

**Comment by Jocs** [comment]
>It would be cool to allow to click on footnote number in footnotes list in exported document to navigate to the footnote in text.

it seems that the number in front of list is not clickable...because it not an html element, but we can add a click element in future improvement.

**Pull request merged**